pkg/extend/extend.go

```diff
@@ -0,0 +1,229 @@
+
+	f := Fdisk{}
+	if err := json.Unmarshal(data, &f); err != nil {
+		return fmt.Errorf("Unable to unmarshal partition table from sfdisk: %v", err)
+	}
+
+	if len(f.PartitionTable.Partitions) > 1 {
+		log.Printf("Disk %s has more than 1 partition. Skipping", d)
+		return nil
+	}
+
+	partition := f.PartitionTable.Partitions[0]
+	if partition.Type != "83" {
+		return fmt.Errorf("Partition 1 on disk %s is not a Linux Partition", d)
+	}
+
+	if partition.Start+partition.Size == f.PartitionTable.LastLBA {
+		log.Printf("No free space on device to extend partition")
+		return nil
+	}
+
+	if err := fsck(partition.Node, false); err != nil {
+		return fmt.Errorf("Initial fsck failed: %v", err)
+	}
+
+	if err := exec.Command("sfdisk", "-q", "--delete", d).Run(); err != nil {
+		return fmt.Errorf("Error erasing partition table: %v", err.Error())
+	}
+
+	createCmd := exec.Command("sfdisk", "-q", d)
+	createCmd.Stdin = strings.NewReader(fmt.Sprintf("%d,,83;", partition.Start))
+	if err := createCmd.Run(); err != nil {
+		return fmt.Errorf("Error creating partition table: %v", err)
+	}
+
+	if err := exec.Command("sfdisk", "-A", d, "1").Run(); err != nil {
+		return fmt.Errorf("Error making %s bootable: %v", d, err)
+	}
+
+	// update status
+	if err := exec.Command("blockdev", "--rereadpt", d).Run(); err != nil {
+		return fmt.Errorf("Error running blockdev: %v", err)
+	}
+
+	exec.Command("mdev", "-s").Run()
+
+	// wait for device
+	var done bool
+	for i := 0; i < timeout; i++ {
+		stat, err := os.Stat(partition.Node)
+		if err == nil {
+			mode := stat.Sys().(*syscall.Stat_t).Mode
+			if (mode & syscall.S_IFMT) == syscall.S_IFBLK {
+				done = true
+				break
+			}
+		}
+		time.Sleep(100 * time.Millisecond)
+		exec.Command("mdev", "-s").Run()
+	}
+	if !done {
+		return fmt.Errorf("Error waiting for device %s", partition.Node)
+	}
+	// even after the device appears we still have a race
+	time.Sleep(1 * time.Second)
+
+	// resize2fs fails unless we set force=true here
+	if err := fsck(partition.Node, true); err != nil {
+		return fmt.Errorf("fsck before resize failed: %v", err)
+	}
+
+	if err := exec.Command("resize2fs", partition.Node).Run(); err != nil {
+		return fmt.Errorf("Error running resize2fs: %v", err)
+	}
+
+	if err := fsck(partition.Node, false); err != nil {
+		return fmt.Errorf("fsck after resize failed: %v", err)
+	}
+
+	log.Printf("Successfully resized %s", d)
+	return nil
+}
+
+func fsck(d string, force bool) error {
+	// preen
+	args := []string{"-p"}
+	if force {
+		args = append(args, "-f")
+	}
+	args = append(args, d)
+	if err := exec.Command("e2fsck", args...).Run(); err != nil {
+		if exiterr, ok := err.(*exec.ExitError); ok {
+			status, ok := exiterr.Sys().(syscall.WaitStatus)
+			if !ok {
+				return fmt.Errorf("Unable to get status code from e2fsck")
+			}
+			switch status.ExitStatus() {
+			case 1:
+				return nil
+			case 2, 3:
+				log.Printf("fsck fixed errors but requires a reboot. rebooting")
+				exec.Command("reboot").Run()
```

Comment:
Not sure we should reboot really. Actually not sure when this might happen either.